File: template/src/com/trip/Application.java
package com.trip;

abstract public class Application {
    private void secure() {
        optimizeAssets();
        useLoadBalancer();
    }

    protected void optimizeAssets() {
        System.out.println("optimizing assets");
    }

    protected void useLoadBalancer() {
        System.out.println("using load balancer");
    }

    private void execute() {
        System.out.println("default execution");
    }

    protected void setupFirewall() {
        System.out.println("setuping firewall");
    }

    protected void enableAttackWatcher() {
        System.out.println("enable attack watcher");
    }

    protected void encryptCommunication() {
        System.out.println("encrypting communication");
    }

    protected void performSecurityScan() {
        System.out.println("performing security scans");
    }

    public void start() {
        secure();
        execute();
    }
}

File: StrategyTemplateMixin/mixup/src/template/Application.scala
package template

abstract class Application {
  private def secure(): Unit = {
    encryptCommunication()
    performSecurityScan()
    enableAttackWatcher()
    setupFirewall()
  }

  private def execute(): Unit = {
    optimizeAssets()
    useLoadBalancers()
  }

  def start(): Unit = {
    secure()
    execute()
  }

  def optimizeAssets(): Unit = println("optimizing assets")
  def useLoadBalancers(): Unit = println("start balancers")


  def setupFirewall(): Unit = println("setup firewall")
  def encryptCommunication(): Unit = println("encrypt communication")
  def performSecurityScan(): Unit = println("apply security")
  def enableAttackWatcher(): Unit = println("enable attack watcher")
}
